code/datums/chatmessage.dm

```diff
@@ -168,69 +230,145 @@
 
 	// Approximate text height
 	var/complete_text = "<span class='center [extra_classes.Join(" ")]' style='color: [tgt_color]'>[owner.say_emphasis(text)]</span>"
-	var/mheight = WXH_TO_HEIGHT(owned_by.MeasureText(complete_text, null, CHAT_MESSAGE_WIDTH))
-	approx_lines = max(1, mheight / CHAT_MESSAGE_APPROX_LHEIGHT)
+	var/measurement = owned_by.MeasureText(complete_text, null, CHAT_MESSAGE_WIDTH)//resolving it to a var so the macro doesnt call MeasureText() twice
+
+	//fun fact: MeasureText() works by waiting for the client to send back the measurements for the text. meaning it works like a verb.
+	//procs like this are called at the last section of hte tick before it ends, meaning if the other portions used up most of it,
+	//then everything after this point is likely to overtime. queuing the message completion if the server is overloaded fixes this
+	if(!owned_by || QDELETED(src) || QDELETED(message_loc) || !all_hearers?[owned_by])
+		return //we should already have been qdel'd() if this evaluates to TRUE, doing it now would throw an error
+
+	var/mheight = WXH_TO_HEIGHT(measurement)
+	if(TICK_CHECK)
+		SSrunechat.message_queue += CALLBACK(src, .proc/generate_image, target, owner, complete_text, mheight)
```

Comment:
we need to use our height when moving older messages upwards

Reply:
I mean more why are we calling WXH_TO_HEIGHT here and not inside generate_image

Reply:
the reason why generate_image() and prepare_text() are two different procs now is because of MeasureText(), the measurement is needed for further processing but reading it puts us into the verb stage of the tick when we're done. if it were in generate_image() we'd have done nothing other than introduce another proc call

Reply:
since we have SSspeech_controller, before we call MeasureText() we should be either 1. the verb stage of the tick BUT the server isnt overloaded, or 2. near the beginning of the tick since the server was overloaded when the client's say() input reached the server so it was queued for the next tick. 

if we dont queue after each client input-waiting proc in a line of processing, we will be putting processing in the potentially dangerous verb portion of the tick where the server may or may not be overloaded, and we're doing stuff at the end of the tick.

Reply:
I'm not saying anything about measure text.
WXH_TO_HEIGHT does not yield, or do anything funky. it just splits some text, and hands back a number.
Why are we doing that splitting before the tick check, rather then inside generate_image.
Also what do you mean by reading it?

Reply:
ah it got refactored while we were talking, nvm potentially, minor overhead but it's a text op and a text2num, not that bad